rfd/9999-aws-api-integration-using-oidc.md

```diff
@@ -0,0 +1,554 @@
+  - `iss`: the issuer identity, on our case it will be the same as the `issuer` key.
+  - `sub`: identifies the subject of the token, in our case it will contain the Teleport username prefixed by `user:`
+  - `aud`: the audience for the token, on our case it will be the same as the `audience` defined when configuring the OIDC IdP in AWS (step 3.3 for the User's Point of View).
+  - `jti`: is the token id (UUIDv4).
+  - `iat`: unix epoch time of when the token was issued.
+  - `exp`: unix epoch time of when the token is no longer valid (expired).
+  - `nbf`: unix epoch time which indicates when the token began to be valid.
+- ID Token Signing Algorithm Values Supported: a list of supported signing algorithms to assist during token validation.
+- Scopes Supported: a list of scopes to be used, on our case it will only contain the required one: `openid`.
+- Subject Types Supported: a list of supported subject types:
+  - `public`: we can either return clear text version of the subject (user names) or
+  - `pairwise`: we can return a hash of the subject's identity (to hide their email, for instance)
+
+##### JSON Web Key Sets
+This endpoint returns a list of public keys (usually only one key, except for periods of key rotation).
+
+This endpoint URI must be equal to `jwks_uri` defined in the previous section.
+
+As an example, `https://proxy.example.com/.well-known/jwks-oidc`.
+
+It should return the following:
+
+```
+200 OK
+
+{
+  "keys": [
+    {
+      "kty": "RSA",
+      "alg": "RS256",
+      "n": "<public key value>",
+      "e": "AQAB"
+    }
+  ]
+}
+```
+
+This must respect the RFC7517 (eg of a key can be found at https://www.rfc-editor.org/rfc/rfc7517#appendix-A.1).
+
+Obtaining this key should be possible using
+```golang
+GetCertAuthority(
+	ctx,
+	types.CertAuthID{
+		Type:       types.OIDCSigner, // new key type
+		DomainName: clusterName,
+	},
+	false /*loadKeys*/,
+)
+```
+
+#### AWS Role for Teleport OIDC IdP
+
+The user will create or associate an AWS Role to the Teleport OIDC IdP.
+
+As an example, we the following is the required policy to list RDS DBs (including Aurora clusters):
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Actions": "[
+				rds:DescribeDBClusters",
+				rds:DescribeDBInstances",
+	        ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+This AWS Role must trust Teleport as an IdP.
+To do so, the user must add the following Trusted relationship:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Federated": "arn:aws:iam::0123456789012:oidc-provider/proxy.example.com"
+            },
+            "Action": "sts:AssumeRoleWithWebIdentity",
+            "Condition": {
+                "StringEquals": {
+                    "proxy.example.com:aud": "sts.amazonaws.com"
```

Comment:
If we used an ID token subject that was specific to the Teleport Discovery process then we could specify this here to ensure that this role can only be used by that process and not by other entities which have been issued tokens.

Reply:
Something like this?
```json
            "Condition": {
                "StringEquals": {
                    "proxy.example.com:aud": "sts.amazonaws.com",
                    "proxy.example.com:sub": "system:proxy:discover",
```

Assuming we issue tokens whose `sub` is `system:proxy:discover`.

What would be the pros of this approach?
Allow for multiple roles to be associated with the same provider but having different policies.

If we are strict here, then we would need users to create extra roles in order to have multiple Teleport "systems" interacting with it.

Honestly, I think we should not add that constraint to not increase scope.
We can always add that condition later on if needed.

Reply:
I'd rather avoid this as it requires further setup to the trust relationship (which is created automatically in the IdP setup flow) with not much added benefit

Reply:
Mhm, I can definitely see the concern from the UX point of view. 

But, I really think as part of the RFD we need to acknowledge that we are opening up access to list all RDS DBs to any machine or user in future if an operator misses an update advisory and does not modify their existing trust relationships created before we issued the change in advice. Whilst this is certainly a low privilege action, one concern I have is that the design from this RfD is used with another feature that requires more privileged permissions and because this wasn't openly acknowledged in this RfD, that this will be missed.

Reply:
We can add
```
 "StringEquals": {
                    "proxy.example.com:aud": "sts.amazonaws.com",
                    "proxy.example.com:sub": "system:proxy",
```

And if we ever need to evolve that, we can just add another suffix to the `sub`.
What do you think?

Reply:
@strideynet I might be misinterpreting your concern, let's chat on Slack if I'm off -

> But, I really think as part of the RFD we need to acknowledge that we are opening up access to list all RDS DBs to any machine or user in future if an operator misses an update advisory and does not modify their existing trust relationships created before we issued the change in advice

Hmm, I don't see this happening - the AWS provider is a one-time setup with an IAM policy, that grants the Teleport auth service access to AWS, which can only be accessed through the AWS integration. This mechanism would be the only thing that sets the `sub` when signing the OIDC JWT, so I don't see the advantage in adding a condition that's always going to evaluate to true

> Whilst this is certainly a low privilege action, one concern I have is that the design from this RfD is used with another feature that requires more privileged permissions and because this wasn't openly acknowledged in this RfD, that this will be missed.

Like above, there's only one feature that's going to use this

Reply:
@marcoandredinis - @strideynet and I had a chat last night about this section and came to the compromise of setting the audience to something other than `sts.amazonaws.com` - maybe `discover.teleport` or `discover.proxy-url`

As per the [AWS docs](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-console) the audience is meant to be the client ID of the application, so setting it to something Discover related is more suitable anyway.

This means we can avoid setting a subject as we're not going to conflict with any future developments of the IdP provider (they'll be requesting a different audience).